apps/api/src/api/services/quote/engines/discount/helpers.ts

```diff
@@ -49,33 +73,86 @@ export async function resolveDiscountPartner(ctx: QuoteContext, rampType: RampDi
  * @param oraclePrice - The oracle price (FIAT-USD format, e.g., BRL-USD)
  * @param targetDiscount - The target discount rate to apply
  * @param isOfframp - Whether this is an offramp (requires price inversion)
+ * @param partnerId - Partner ID for state management
  * @returns Expected output amount
  */
 export function calculateExpectedOutput(
   inputAmount: string,
   oraclePrice: Big,
   targetDiscount: number,
-  isOfframp: boolean
+  isOfframp: boolean,
+  partner: ActivePartner
 ): Big {
   const inputAmountBig = new Big(inputAmount);
 
   // For offramps, we need to invert the oracle price
   // Oracle price is FIAT-USD, so for offramps we want USD-FIAT
   const effectivePrice = isOfframp ? new Big(1).div(oraclePrice) : oraclePrice;
 
-  // Apply target discount to the rate
-  const discountedRate = effectivePrice.mul(new Big(1).plus(targetDiscount));
+  // Apply target discount to the rate, adjusting first for dynamic discount variable.
+  const adjustedTargetDiscount = new Big(targetDiscount).plus(getAdjustedDifference(partner));
+  const discountedRate = effectivePrice.mul(new Big(1).plus(adjustedTargetDiscount));
   return inputAmountBig.mul(discountedRate);
 }
 
-/**
- * Calculate the subsidy amount by comparing expected vs actual output
- * Caps at maxSubsidy if configured
- * @param expectedOutput - The expected output amount
- * @param actualOutput - The actual output amount from nabla swap
- * @param maxSubsidy - Maximum subsidy rate (decimal)
- * @returns Subsidy amount as Big
- */
+export function getAdjustedDifference(partner?: ActivePartner): Big {
+  if (!partner?.id) {
+    return new Big(0);
+  }
+
+  const partnerState = partnerDiscountState.get(partner.id);
+  const now = new Date();
+
+  // Use partner's min caps if available, otherwise fall back to constants
+  const maxCap = partner.maxTargetDiscount ?? MAX_DIFFERENCE_CAP;
```

Comment:
I think we shouldn't fall back to the constants if it's not available. We should rather fall back to the `partner.targetDiscount` field if available.